sympy/core/basic.py

```diff
@@ -1558,7 +1498,7 @@ def _has(self, iterargs, *patterns):
         # no success
         return False
 
-    def replace(self, query, value, map=False, simultaneous=True, exact=None):
+    def replace(self, query, value, map=False, simultaneous=True, exact=None) -> Basic:
```

Comment:
I'm not sure about forcing this to Basic because `map` definitely chages the return type
```
x = Symbol('x')
y = Symbol('y')
f = Function('f')
print(f(x).replace(x, y, map=True))
print(f(x).replace(x, y, map=False))
```
```
(f(y), {x: y})
f(y)
```

Reply:
The map parameter is not used anywhere:
```console
$ git grep 'map=True'
sympy/core/basic.py:            >>> sin(x).replace(sin, cos, map=True)
sympy/core/tests/test_expr.py:    assert cos(x).replace(cos, sin, map=True) == (sin(x), {cos(x): sin(x)})
sympy/core/tests/test_expr.py:    assert (x*y).replace(cond, func, map=True) == (2*x*y, {x*y: 2*x*y})
sympy/core/tests/test_expr.py:    assert (x*(1 + x*y)).replace(cond, func, map=True) == \
sympy/core/tests/test_expr.py:    assert (y*sin(x)).replace(sin, lambda expr: sin(expr)/y, map=True) == \
sympy/core/tests/test_expr.py:    assert e.replace(cond, func, map=True) == (
```
I would rather not make a complicated overload signature here for something that is never used (and should have been a separate function anyway).

Reply:
Does that mean that the return type will be Tuple(Expr, Dict) if map is used?

Reply:
The type hint says that the return type is always `Basic` which is true everywhere the method is called within the codebase apart from a couple of tests for `map=True`. That means that if a type checker tries to infer types from any code that calls replace it will infer the correct type Basic.

The replace method has a very complicated signature and expressing fully what it returns probably requires many overloads. The `map=True` parameter would double the number of overloads so I would rather overlook it right now.

I think that @sylee957 dislikes the fact that the type hint is not accurate when `map=True` is passed but right now the most important thing is just getting the types to be inferred correctly most of the time. The real problem here is that usually `Expr.replace` returns `Expr` (rather than `Basic`) but the type hint here does not express that fact. We would need at least separate overloads for `Expr` and `Basic` here to get basic inference working and then typing all the parameters fully would give at least 4 overloads so that's already `2*4=8` overloads and then `map=True` doubles everything giving at least 16 overloads.

Reply:
> I think that @sylee957 dislikes the fact that the type hint is not accurate when map=True is passed

I don't think that it's about 'dislike', but that is imprecise way of writing type annotations and that's why it could have been questioned in review by me or anyone else to notice.

I'm okay if the typing is kept as Basic,
However, my opinion is that it could be better to use deprecations for such things.

Reply:
We could also deprecate it. I think that having replace be so overloaded is a mistake and at least the pattern matching case should be a separate method because it has a very different signature and the `map=True` case should not be there at all.

For now it is not really possible for the type hints to be precise because it is still unclear (to me at least) what the correct precise type hints would be in most cases. I think that we may end up needing a type parameter like `Basic[T]` where `T` refers to the same sort of thing as `.kind` e.g. to distinguish matrices from numbers etc. The actual type like `Integral` does not really represent the "type" correctly:
```
In [21]: i = Integral(Matrix([[1, 2], [3, 4]]), (x, 0, 1))

In [22]: i
Out[22]: 
1          
⌠          
⎮ ⎡1  2⎤   
⎮ ⎢    ⎥ dx
⎮ ⎣3  4⎦   
⌡          
0          

In [23]: shape(i)
Out[23]: (2, 2)

In [24]: i.kind
Out[24]: MatrixKind(NumberKind)

In [25]: type(i)
Out[25]: sympy.integrals.integrals.Integral
```
Probably a type parameter like this is also needed to be able to express the fact that `Expr.subs` always returns `Expr`.